pkg/controllers/upgrade_reconcile.go

```diff
@@ -0,0 +1,271 @@
+			"Vertica server upgrade has started.  Upgrading to image '%s'", u.Vdb.Spec.Image)
+	}
+	return ctrl.Result{}, nil
+}
+
+// finishUpgrade handles condition status and event recording for the end of an upgrade
+func (u *UpgradeReconciler) finishUpgrade(ctx context.Context) (ctrl.Result, error) {
+	if err := u.toggleUpgradeInProgress(ctx, corev1.ConditionFalse); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	u.VRec.EVRec.Eventf(u.Vdb, corev1.EventTypeNormal, events.UpgradeSucceeded,
+		"Vertica server upgrade has completed successfully")
+
+	return ctrl.Result{}, nil
+}
+
+// isUpgradeNeeded returns true if we are in the middle of an upgrade or we need to start one
+func (u *UpgradeReconciler) isUpgradeNeeded(ctx context.Context) (bool, error) {
+	// We first check if the status condition indicates the upgrade is in progress
+	inx, ok := vapi.VerticaDBConditionIndexMap[vapi.UpgradeInProgress]
+	if !ok {
+		return false, fmt.Errorf("verticaDB condition '%s' missing from VerticaDBConditionType", vapi.UpgradeInProgress)
+	}
+	if inx < len(u.Vdb.Status.Conditions) && u.Vdb.Status.Conditions[inx].Status == corev1.ConditionTrue {
+		// Set a flag to indicate that we are continuing an upgrade.  This silences the UpgradeStarted event.
+		u.ContinuingUpgrade = true
+		return true, nil
+	}
+
+	// Next check if an upgrade is needed based on the image being different
+	// between the Vdb and any of the statefulset's.
+	finder := MakeSubclusterFinder(u.VRec.Client, u.Vdb)
+	stss, err := finder.FindStatefulSets(ctx, FindInVdb)
+	if err != nil {
+		return false, err
+	}
+	for i := range stss.Items {
+		sts := stss.Items[i]
+		if sts.Spec.Template.Spec.Containers[names.ServerContainerIndex].Image != u.Vdb.Spec.Image {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// toggleUpgradeInProgress is a helper for updating the UpgradeInProgress condition
+func (u *UpgradeReconciler) toggleUpgradeInProgress(ctx context.Context, newVal corev1.ConditionStatus) error {
+	return status.UpdateCondition(ctx, u.VRec.Client, u.Vdb,
+		vapi.VerticaDBCondition{Type: vapi.UpgradeInProgress, Status: newVal},
+	)
+}
+
+// stopCluster will shutdown the entire cluster using 'admintools -t stop_db'
+func (u *UpgradeReconciler) stopCluster(ctx context.Context) (ctrl.Result, error) {
+	pf, found := u.PFacts.findRunningPod()
+	if !found {
+		u.Log.Info("No pods running so skipping vertica shutdown")
+		// No running pod.  This isn't an error, it just means no vertica is
+		// running so nothing to shut down.
+		return ctrl.Result{}, nil
+	}
+	if u.PFacts.getUpNodeCount() == 0 {
+		u.Log.Info("No vertica process running so nothing to shutdown")
+		// No pods have vertica so we can avoid stop_db call
+		return ctrl.Result{}, nil
+	}
+
+	// Check the image of each of the up nodes.  We avoid doing a shutdown if
+	// all of the up nodes are already on the new image.
+	if ok, err := u.anyPodsRunningWithOldImage(ctx); !ok || err != nil {
```

Comment:
Trying to understand the process: am I understanding correctly that there could be a period where Pods with old images and Pods with new images co-exist, it's just in such a period we make sure the DB is in shutdown status.

Reply:
This is here to see whether we need to do a shutdown.  The entire reconciler is idempotent, so it is just checking to see if a shutdown is necessary.  There are three cases we are considering here:
1. database is up with the old image
2. database is already shutdown
3. the cluster is already up with the new image

We skip the shutdown for case 2 and 3.

I'll update the comments here.